### PR TITLE
BOLEClassifier + DriftRetrainingClassifier

### PR DESCRIPTION
**Pull request by danielnowakassis**

### CODE REVIEW

File: river/ensemble/boosting.py [hunk 1]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+        for i in range(len(models)):
+            if correct:
+                pos = self.orderPosition[maxAcc]
+                maxAcc += 1
+            else:
+                pos = self.orderPosition[minAcc]
+                minAcc -= 1
+
+            for _ in range(poisson(lambda_poisson, self._rng)):
+                models[pos][1].learn_one(x, y)
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:238]: You can do `self[pos][1].learn_one`, which avoids `models = list(enumerate(self))`

File: river/ensemble/boosting.py [hunk 2]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+
+    def predict_proba_one(self, x):
+        y_proba = collections.Counter()
+        for i, model in enumerate(self):
+            model_weight = 0.0
+            if self.correct_weight[i] > 0.0 and self.wrong_weight[i] > 0.0:
+                epsilon = self.wrong_weight[i] / (self.correct_weight[i] + self.wrong_weight[i])
+                if epsilon <= self.error_bound:
+                    beta_inv = (1 - epsilon) / epsilon
+                    model_weight = np.log(beta_inv)
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:258]:
```suggestion
                    model_weight = math.log(beta_inv)
```

File: river/ensemble/boosting.py [hunk 3]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+                predictions = model.predict_proba_one(x)
+                normalize_values_in_dict(predictions, inplace=True)
+                scale_values_in_dict(predictions, model_weight, inplace=True)
+                y_proba.update(predictions)
+
+        normalize_values_in_dict(y_proba, inplace=True)
+        return y_proba
+
+
+class BOLEBaseModel(base.Classifier):
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:270]: I really like this class. In fact, I think it should be a class of its own that would live in the `drift` module. It's really useful I would call it something more generic, such as `RetrainOnDriftClassifier`

File: river/ensemble/boosting.py [hunk 4]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+
+    Parameters
+    ----------
+    model
+        The classifier and background classifier class.
+    drift_detection_method
+        Algorithm to track warnings and concept drifts
+    """
+
+    def __init__(self, model: base.Classifier, drift_detection_method: DriftDetector):
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:283]:
```suggestion
    def __init__(self, model: base.Classifier, drift_detector: DriftDetector = None):
```

File: river/ensemble/boosting.py [hunk 5]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+        self.base_class = model
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:284]: One thing you can do is only store the `model` attribute. Then, you can call `model.clone()` to get reinitialize the model, without having to store `base_class`

File: river/ensemble/boosting.py [hunk 6]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+    def predict_proba_one(self, x):
+        return self.model.predict_proba_one(x)
+
+    def learn_one(self, x, y):
+        self.update_ddm(x, y)
+        return self.model.learn_one(x, y)
+
+    def update_ddm(self, x, y):
+        y_pred = self.model.predict_one(x)
+        if y_pred is not None:
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:300]: It's preferable to do 

```py
if y_pred is None:
    return
```

and then continue. It's more readable that way

File: river/ensemble/boosting.py [hunk 7]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+            incorrectly_classifies = int(y_pred != y)
+            self.drift_detection_method.update(incorrectly_classifies)
+            if self.drift_detection_method.warning_detected:
+                if self.bkg_model is None:
+                    self.bkg_model = copy.deepcopy(self.base_class)
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:305]: You can use `.clone()` instead of `copy.deepcopy`

File: river/ensemble/boosting.py [hunk 8]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+        y_proba = collections.Counter()
+        for i, model in enumerate(self):
+            model_weight = 0.0
+            if self.correct_weight[i] > 0.0 and self.wrong_weight[i] > 0.0:
+                epsilon = self.wrong_weight[i] / (self.correct_weight[i] + self.wrong_weight[i])
+                if epsilon <= self.error_bound:
+                    beta_inv = (1 - epsilon) / epsilon
+                    model_weight = np.log(beta_inv)
+
+            if model_weight != 0.0:
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:260]:
```suggestion
            if model_weight:
```

File: river/ensemble/boosting.py [hunk 9]
```diff
@@ -1,12 +1,17 @@
 import collections
+import copy
 import math
 import typing
 
+import numpy as np
+
 from river import base, linear_model
+from river.base import DriftDetector
+from river.drift import DDM
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:10]: Avoid using absolute imports. It's more Pythonic to import `base` and `drift` and then access the classes you want.

File: river/ensemble/boosting.py [hunk 10]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+    ...     seed=42
+    ... )
+
+    >>> metric = metrics.Accuracy()
+
+    >>> evaluate.progressive_val_score(dataset, model, metric)
+    Accuracy: 92.02%
+
+
+
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:185]:
```suggestion

```

File: river/ensemble/boosting.py (hunk 7 above)
Comment (MaxHalford) [inline river/ensemble/boosting.py:305]: I think you can do `self.bkg_model = copy.deepcopy(self.base_class)` (or `self.base_class.clone()`) in the `__init__` (on line 286) and then delete this `if`. Then you can call `self.bkg_model = copy.deepcopy(self.base_class)` (or `self.base_class.clone()`) at line 310

Comment (MaxHalford) [reply river/ensemble/boosting.py:305]: But maybe I'm wrong! I'll let you check :)

**Review by MaxHalford — CHANGES_REQUESTED** [review]
Hey! I made some comments. Overall I'm very pleased with the simplicity of the method, as well as the model's performance (apparently).

I really like the `BOLEBaseModel` class, and I think it should be a class of its own. People will want to use it.

There are some code style issues. I've commented most of them. Don't feel negative about them, we're just a bit maniac in our codebase :). Also, if it helps I can make a pull request on your pull request to suggest some changes.

Cheers!

File: river/ensemble/boosting.py [hunk 11]
```diff
@@ -119,3 +124,292 @@ def predict_proba_one(self, x):
...
+    Data Mining, 2009.
+
+    [^2]: Oza, N., Russell, S. "Online bagging and boosting."
+    In: Artificial Intelligence and Statistics 2001, pp. 105–112.
+    Morgan Kaufmann, 2001.
+
+    """
+    def __init__(self, model: base.Classifier, n_models=10, seed: int = None):
+        super().__init__(model, n_models, seed)
+        self._drift_detectors = [copy.deepcopy(ADWIN()) for _ in range(self.n_models)]
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:188]: Would this model work with other drift detectors? In theory? Does it have to just for `ADWIN`?

File: river/ensemble/boosting.py [hunk 12]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+            j = i - 1
+            while (j >= 0) and (acc[j] < key_acc):
+                self.orderPosition[j + 1] = self.orderPosition[j]
+                acc[j + 1] = acc[j]
+                j -= 1
+            self.orderPosition[j + 1] = key_position
+            acc[j + 1] = key_acc
+
+        correct = False
+        maxAcc = 0
```
[the remark this answers is not in the document]
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: resolved in snake_case + range() commit

File: river/ensemble/boosting.py [hunk 13]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+    ----------
+    [^1]: [Oza, N.C., 2005, October. Online bagging and boosting. In 2005 IEEE international conference on systems, man and cybernetics (Vol. 3, pp. 2340-2345). Ieee.](https://ti.arc.nasa.gov/m/profile/oza/files/ozru01a.pdf)
+
+    [^2]: R. S. M. d. Barros, S. Garrido T. de Carvalho Santos and P. M. Gonçalves Júnior, "A Boosting-like Online Learning Ensemble," 2016 International Joint Conference on Neural Networks (IJCNN), 2016, pp. 1871-1878, doi: 10.1109/IJCNN.2016.7727427.
+    """
+
+    def __init__(self, model: base.Classifier, n_models=10, seed: int = None, error_bound=0.5):
+        super().__init__(model=model, n_models=n_models, seed=seed)
+        self.error_bound = error_bound
+        self.orderPosition = np.arange(n_models)
```
[the remark this answers is not in the document]
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: resolved in snake_case + range() commit. acc in learn_one will be replaced with [0] * 9. If you have a better suggestion, please comment :)

File: river/ensemble/boosting.py (hunk 1 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: the object is not subscriptable. Solved in last numpy removes and self.models[ind] commit. River has a self.models that returns the models.

File: river/ensemble/boosting.py (hunk 2 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: Solved in last numpy removals and self.models[ind] commit.

File: river/ensemble/boosting.py (hunk 3 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py:267]: It´s better to do this in another pull-request.

Comment (danielnowakassis) [reply river/ensemble/boosting.py:267]: I will substitute the changes suggested and then take BOLEBaseModel out

File: river/ensemble/boosting.py (hunk 5 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py:281]: For me it is not clear what you intend. Can you modify?

Comment (danielnowakassis) [reply river/ensemble/boosting.py:281]: For me it is not clear what you want. Can you modify it?

File: river/ensemble/boosting.py (hunk 4 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py:280]: Solved in BaseModel + other changes

File: river/ensemble/boosting.py (hunk 7 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py:302]: I use None to avoid training without having a model (and maybe avoid some bugs).

Comment (danielnowakassis) [reply river/ensemble/boosting.py:302]: What's the difference in this context?

File: river/ensemble/boosting.py [hunk 14]
```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
...
+                models[pos][1].learn_one(x, y)
+
+            if models[pos][1].predict_one(x) == y:
+                self.correct_weight[pos] += lambda_poisson
+                lambda_poisson *= (self.instances_seen) / (2 * self.correct_weight[pos])
+                correct = True
+            else:
+                self.wrong_weight[pos] += lambda_poisson
+                lambda_poisson *= (self.instances_seen) / (2 * self.wrong_weight[pos])
+                correct = False
```
[the remark this answers is not in the document]
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: It's hard to explain, but i will do my best. Tried to solve in BaseModel + other changes

File: river/ensemble/boosting.py (hunk 6 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: Solved in BaseModel + other changes

File: river/ensemble/boosting.py (hunk 8 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: Solved in BaseModel + other changes

File: river/ensemble/boosting.py (hunk 9 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: Solved in BaseModel + other changes

File: river/ensemble/boosting.py (hunk 10 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: Solved in BaseModel + other changes

File: river/ensemble/boosting.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: * self.n_models *

**Comment by MaxHalford** [comment]
Looking good @danielnowakassis! Could you give me push access to your fork? I'll push a few changes if that's ok :)

**Comment by MaxHalford** [comment]
Ok @danielnowakassis I pushed a fair number of changes. I'll let you take a look. What do you think?

**Comment by danielnowakassis** [comment]
I agree with all changes @MaxHalford. We have done some good work :) You can merge it.

**Comment by MaxHalford** [comment]
Agreed! Before merging though, we still need to fix the tests. On my end I'll look into test for `DriftRetrainingClassifier`. Can you check for `BOLEClassifier`?

**Comment by MaxHalford** [comment]
Good job @danielnowakassis!

**Comment by danielnowakassis** [comment]
Thanks Max. We have done some good work 🙂

**Pull request merged**